File: solutions/typescript/arrays-and-hashing/product-of-array-except-self.ts
/**
 * 238. Product of Array Except Self — https://leetcode.com/problems/product-of-array-except-self/
 *
 * Each output position holds the product of every input element except the one beneath it.
 *
 * Division is forbidden, and that restriction is doing real work: "total ÷ nums[i]" breaks the
 * moment a zero appears, and breaks differently for one zero than for two.
 */

import { defineProblem } from "../define-problem";

export const SLUG = "product-of-array-except-self";

/**
 * Collapse `-0` to `0`.
 *
 * IEEE 754 has a signed zero and JavaScript exposes it. For `[-1, 1, 0, -3, 3]` the product for
 * index 0 is `1 * 0 * -3 * 3`, which evaluates left to right as `0 → -0 → -0`. Python's and
 * PHP's integers have no signed zero at all, so both produce a plain `0` and this
 * implementation was the odd one out — caught by the shared contract's "contains one zero"
 * case, which is exactly what it is there for.
 *
 * Worth knowing how narrowly this escapes notice: `-0 === 0` is `true`, so no ordinary
 * comparison sees it, and `JSON.stringify(-0)` is `"0"`, so the HTTP response would have agreed
 * with the other two apps regardless. Only a structural deep-equal — which is what the test
 * does, and what `Object.is` does — can tell them apart.
 *
 * `value === 0` matches both zeros, so returning the literal normalises either to `+0`.
 */
function normalizeZero(value: number): number {
  return value === 0 ? 0 : value;
}

/**
 * Recompute each product from scratch. Time O(n^2), space O(1) beyond the output.
 *
 * The empty product is 1, which is why a single-element input returns `[1]`.
 */
export function productOfArrayExceptSelfBruteForce(nums: readonly number[]): number[] {
  const result: number[] = [];
  for (let i = 0; i < nums.length; i++) {
    let product = 1;
    for (let j = 0; j < nums.length; j++) {
      if (i !== j) product *= nums[j]!;
    }
    result.push(normalizeZero(product));
  }
  return result;
}

/**
 * Left products, then fold the right products in. Time O(n), space O(1) beyond the output.
 *
 * The answer at `i` factorises into everything left of `i` times everything right of `i`. The
 * first pass writes the left half into the output array; the second walks backwards carrying
 * the right half in a single variable, so no second array is ever allocated.
 *
 * `new Array(n).fill(1)` rather than `new Array(n)`: the latter creates a *sparse* array whose
 * holes are `undefined`, and `undefined *= x` yields `NaN`. Python's `[1] * n` and PHP's
 * `array_fill` have no equivalent hazard.
 */
export function productOfArrayExceptSelfPrefixSuffix(nums: readonly number[]): number[] {
  const result = new Array<number>(nums.length).fill(1);

  let prefix = 1;
  for (let i = 0; i < nums.length; i++) {
    result[i] = prefix;
    prefix *= nums[i]!;
  }

  let suffix = 1;
  for (let i = nums.length - 1; i >= 0; i--) {
    result[i] = normalizeZero(result[i]! * suffix);
    suffix *= nums[i]!;
  }

  return result;
}

export const productOfArrayExceptSelf = defineProblem(SLUG, {
  "brute-force": productOfArrayExceptSelfBruteForce,
  "prefix-suffix": productOfArrayExceptSelfPrefixSuffix,
});

File: solutions/python/arrays_and_hashing/product_of_array_except_self.py
"""238. Product of Array Except Self — https://leetcode.com/problems/product-of-array-except-self/

Each output position holds the product of every input element *except* the one beneath it.

Division is forbidden by the problem, and that restriction is doing real work: the obvious
"total product ÷ nums[i]" breaks the moment a zero appears, and breaks differently for one zero
than for two.
"""

from __future__ import annotations

from neetcode_core.registry import solution

SLUG = "product-of-array-except-self"


@solution(SLUG, approach="brute-force")
def product_except_self_brute_force(nums: list[int]) -> list[int]:
    """Recompute each product from scratch. Time O(n^2), space O(1) beyond the output.

    The empty product is 1, which is why a single-element input returns `[1]` rather than
    `[0]` or an error. Worth stating out loud — it is the case every rewrite gets wrong.
    """
    result = []
    for i in range(len(nums)):
        product = 1
        for j, value in enumerate(nums):
            if i != j:
                product *= value
        result.append(product)
    return result


@solution(SLUG, approach="prefix-suffix")
def product_except_self_prefix_suffix(nums: list[int]) -> list[int]:
    """Left products, then fold the right products in. Time O(n), space O(1) beyond the output.

    The insight is that the answer at `i` factorises into everything left of `i` times
    everything right of `i`. The first pass writes the left half into the output array; the
    second walks backwards carrying the right half in a single variable, so no second array is
    ever allocated.

    The output array does not count toward the space bound — that is the problem's own
    convention, and it is the only reason this counts as O(1).
    """
    result = [1] * len(nums)

    prefix = 1
    for i, value in enumerate(nums):
        result[i] = prefix
        prefix *= value

    suffix = 1
    for i in range(len(nums) - 1, -1, -1):
        result[i] *= suffix
        suffix *= nums[i]

    return result
